osu.Game.Rulesets.Osu/Mods/OsuModAutopilot.cs

```diff
@@ -36,21 +39,63 @@ public class OsuModAutopilot : Mod, IApplicableFailOverride, IUpdatableByPlayfie
 
         private int currentFrame;
 
+        int cachedFrames = 1;
+        private const float relax_leniency = 3f;
+
         public void Update(Playfield playfield)
         {
             if (currentFrame == replayFrames.Count - 1) return;
 
-            double time = gameplayClock.CurrentTime;
+            double time = playfield.Clock.CurrentTime;
+
+            bool breakNow = false;
+            bool pause = false;
+            foreach (var h in playfield.HitObjectContainer.AliveObjects.OfType<DrawableOsuHitObject>())
+            {
+                // We are not yet close enough to the object time
+                if(time < h.HitObject.StartTime - relax_leniency) 
+                    break;
+
+                // Already hit or beyond the hittable end time.
+                // Here a relax_leniency is used instead of end time because end time causes lags for some reason
+                if(h.IsHit || time > h.HitObject.StartTime + relax_leniency)
+                    continue;
+
+                switch (h)
+                {
+                    case DrawableHitCircle circle:
+                        pause = true;
+                        breakNow = true;
+                        break;
+
+                    case DrawableSlider slider:
+                        breakNow = true;
+                        break;
+
+                    case DrawableSpinner _:
+                        breakNow = true;
+                        break;
+                }
+
+                if(breakNow) {
+                    break;
+                }
+            }
 
-            // Very naive implementation of autopilot based on proximity to replay frames.
-            // TODO: this needs to be based on user interactions to better match stable (pausing until judgement is registered).
-            if (Math.Abs(replayFrames[currentFrame + 1].Time - time) <= Math.Abs(replayFrames[currentFrame].Time - time))
+            // If it's time to move the cursor, check if it needs to be paused 
+            // (to account for player error when tapping)
+            // If not, catch up to the current position required
+            if(Math.Abs(replayFrames[currentFrame + cachedFrames].Time - time) <= Math.Abs(replayFrames[currentFrame].Time - time))
```

Comment:
what are "cached frames"? how does this work?

Reply:
I intended for the cached frames to act as a buffer to the frames that the cursor has to move. Where if the player clicks a bit later than usual it will still register. After the click is registered, the cursor will catch up to where it is supposed to be.